trace/src/main/org/jetbrains/lincheck/trace/Deserialization.kt

```diff
@@ -0,0 +1,771 @@
+            tracePoint.loadFooter(data)
+        } else {
+            System.err.println("TraceRecorder: Unexpected object kind $kind when loading tracepoints")
+        }
+
+        return tracePoint
+    }
+
+    private fun readTracePointWithChildAddresses(): TRTracePoint {
+        // Load tracepoint itself
+        val tracePoint = loadTRTracePoint(data)
+        if (tracePoint !is TRMethodCallTracePoint) {
+            return tracePoint
+        }
+
+        val positions = callTracepointChildren[tracePoint.eventId]
+            ?: error("Tracepoint ${tracePoint.eventId} is not known in index")
+
+        val checkFor = calculatePhysicalOffset(tracePoint.threadId, positions.first)
+        check(data.position() == checkFor) {
+            "TRMethodCallTracePoint ${tracePoint.eventId} has wrong start position in index: ${positions.first} / $checkFor, expected ${data.position()}"
+        }
+
+        // Read tracepoints truly shallow
+        loadTracePoints(tracePoint.threadId, Integer.MAX_VALUE, this::readTracePointShallow) { _, _, address ->
+            tracePoint.addChildAddress(calculateLogicalOffset(tracePoint.threadId, address))
+        }
+        // Kind is guaranteed by loadTracePoints()
+        tracePoint.loadFooter(data)
+
+        return tracePoint
+    }
+
+    private fun calculatePhysicalOffset(threadId: Int, logicalOffset: Long): Long {
+        val blocks = dataBlocks[threadId] ?: error("ThreadId $threadId is not found in block list")
+        val blockIdx = blocks.binarySearch { it.compareWithLogicalOffset(logicalOffset) }
+        check(blockIdx >= 0) { "Thread $threadId doesn't have data at logical offset $logicalOffset" }
+        val block = blocks[blockIdx]
+        return block.physicalDataStart + logicalOffset - block.accDataSize
+    }
+
+    private fun calculateLogicalOffset(threadId: Int, physicalOffset: Long): Long {
+        val blocks = dataBlocks[threadId] ?: error("ThreadId $threadId is not found in block list")
+        val blockIdx = blocks.binarySearch { it.compareWithPhysicalOffset(physicalOffset) }
+        check(blockIdx >= 0) { "Thread $threadId doesn't have data at physical offset $physicalOffset" }
+        val block = blocks[blockIdx]
+        return block.accDataSize + physicalOffset - block.physicalDataStart
+    }
+
+    private fun findBlockByPhysicalOffset(threadId: Int, physicalOffset: Long): Int? {
+        val blocks = dataBlocks[threadId] ?: error("ThreadId $threadId is not found in block list")
+        val blockIdx = blocks.binarySearch { it.compareWithPhysicalOffset(physicalOffset) }
+        return if (blockIdx < 0) null else blockIdx
+    }
+
+    private companion object {
+        fun wrapStream(input: InputStream?): DataInputStream? {
+            if (input == null) return null
+            return DataInputStream(input.buffered(INPUT_BUFFER_SIZE))
+        }
+    }
+}
+
+fun loadRecordedTrace(inp: InputStream): Pair<TraceContext, List<TRTracePoint>> {
```

Comment:
Worth adding a data class with proper field names instead of  `Pair<TraceContext, List<TRTracePoint>>`